conda/core/solve/state.py

```diff
@@ -0,0 +1,717 @@
+
+    @property
+    def with_only_deps(self) -> DepsModifier:
+        return str(self._deps_modifier) == str(DepsModifier.ONLY_DEPS)
+
+    # Other flags
+
+    @property
+    def ignore_pinned(self) -> bool:
+        return self._ignore_pinned
+
+    @property
+    def force_remove(self) -> bool:
+        return self._force_remove
+
+    @property
+    def force_reinstall(self) -> bool:
+        return self._force_reinstall
+
+    @property
+    def prune(self) -> bool:
+        return self._prune
+
+
+class SolverOutputState(Mapping):
+    """
+    This is the main mutable object we will massage before passing the
+    result of the computation to the solver. It will also store the result
+    of the solve.
+
+    Its main intent is to populate the different spec pools with MatchSpec
+    objects obtained from an ``SolverInputState`` instance.
+
+    The main result of this operation is available through ``.pool_items()``,
+    which returns ``str, MatchSpec`` tuples, mimicking ``dict.items()``. The
+    key here is the pool name, which stands for a reason why this spec
+    made it to the final list.
+
+    The pool merging logic is done in ``._populate_pools()``.
+
+    All of these pools will map package names (``str``) to ``MatchSpec`` (_specs_ in short)
+    objects. A quick note on these objects:
+
+    * They are a query language for packages, based on the ``PackageRecord`` schema.
+      ``PackageRecord`` objects is how packages that are already installed are represented. This
+      is what you get from ``PrefixData.iter_records()``. Since they are related,
+      ``MatchSpec`` objects can be created from a ``PackageRecord``.
+    * ``MatchSpec`` objects also feature fields like ``target`` and ``optional``. These are,
+      essentially, used by the low-level classic solver (:class:`conda.resolve.Resolve`) to mark
+      specs as items they can optionally play with to satisfy the solver constrains. A ``target``
+      marked spec is _soft-pinned_ in the sense that the solver will try to satisfy that but it
+      will stop trying if it gets in the way, so you might end up a different version or build.
+      ``optional`` seems to be in the same lines, but maybe the entire spec can be dropped from the
+      request? The key idea here is that these two fields might not be directly usable by the solver,
+      but it might need some custom adaptation. For example, for ``libmamba`` we might need a separate
+      pool that can be configured as a flexible task. See more details in the first comment of
+      ``conda.core.solve.classic.Solver._add_specs``
+    """
+    def __init__(
+        self,
+        *,
+        specs: Optional[Mapping[str, MatchSpec]] = None,
+        records: Optional[Mapping[str, PackageRecord]] = None,
+        neutered: Optional[Mapping[str, MatchSpec]] = None,
+        conflicts: Optional[Mapping[str, MatchSpec]] = None,
+        solver_input_state: Optional[SolverInputState] = None,
```

Comment:
Can we add comments on those?

Reply:
I'll put them in the docstring once we have a definite list of args here.